ci: refactor lint CI task and fix lint issues

RaschidJFR opened the pull request:
### New Pull Request Checklist
<!--
    Please check the following boxes [x] before submitting your issue.
    Click the "Preview" tab for better readability.
    Thanks for contributing to Parse Server!
-->

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-dashboard/issues?q=is%3Aissue).

### Issue Description
<!-- Add a brief description of the issue this PR solves. -->

Currently, the commands `npm test` and `npm run lint` are failing: some files are not following style rules and there are some outdated test cases.

### Approach
<!-- Add a description of the approach in this PR. -->
* Lint-fix all files
* Update outdated test cases to pass
* Add unit testing to CI workflow

### TODOs before merging
<!--
    Add TODOs that need to be completed before merging this PR.
    Delete TODOs that do not apply to this PR.
-->

- [x] Add tests
- [ ] ~~Add changes to documentation (guides, repository pages, in-code descriptions)~~
- [x] A changelog entry is created automatically using the pull request title (do not manually add a changelog entry)

RaschidJFR committed:
chore(vscode): add Jest types

RaschidJFR committed:
style(*): fix lint errors

RaschidJFR committed:
test(*): update failing tests

RaschidJFR committed:
chore(workflows): add step `check-jest`

mtrezza commented on release.config.js:
```diff
@@ -86,7 +86,7 @@ async function config() {
       ['@semantic-release/github', {
         successComment: getReleaseComment(),
         labels: ['type:ci'],
-        releasedLabels: ['state:released<%= nextRelease.channel ? `-\${nextRelease.channel}` : "" %>']
```
Did you manually change this? I think to remember that the `\` was required, because this is not supposed to parsed as a placeholder at this point, but I may be wrong.

RaschidJFR replied:
No, it was the linter. Maybe we should apply linting only to the `src` folder?

mtrezza replied:
Let me verify whether we need the `\`.

mtrezza replied:
It seems we can safely remove the `\`, I tested it and the release labels are correctly applied.

mtrezza commented on src/dashboard/Data/Browser/Browser.react.js:
```diff
@@ -505,7 +505,7 @@ class Browser extends DashboardView {
           if (name === 'objectId' || this.state.isUnique && name !== this.state.uniqueField) {
             return;
           }
-          if (!!required) {
```
I am always suspicious of that rule, because `!!` is not the same as removing it, if there are no type checks.

RaschidJFR replied:
Got you, but I believe this [rule](https://eslint.org/docs/rules/no-extra-boolean-cast) acts only in the cases in which `!!` makes no difference. I can't think of a scenario in which the rule would actually harm the logic.

It's not a problem to disable it, but you know, the less we customize the less we have to maintain.

mtrezza replied:
I think we can leave the rule, I checked and the double bang is only removed within `if` conditions where it's unnecessary. Let's hope the rule is intelligent enough to remove only the unnecessary ones, as you said.

mtrezza commented on .github/workflows/ci.yml:
```diff
@@ -29,26 +29,27 @@ jobs:
+      - name: Install dependencies
+        run: npm ci
+      - name: Test with Jest
+        run: npm test
```
Why are we not using `npm run lint` for this? I think it would be easier, as we want to unify the linting and rules across repos and can use a shared config.

RaschidJFR replied:
What do you mean, adding another name-run with `npm run lint`? The lint command is currently run in [pretest](https://github.com/RaschidJFR/parse-dashboard/blob/bugfix/fix_lint_and_tests/package.json#L130)

mtrezza replied:
I didn't notice that, it seems to be the only repo where lint runs in pretest. I suggest to remove pretest lint entirely, re-add the previously commented-out (now removed) lint CI job with `npm run lint` (that's why I've added it provisionally) and maybe add the jtest test to the matrix job `check-build`, if it makes sense to run the tests under various node versions after building the bundles.

RaschidJFR replied:
I believe that it would make sense to run the e2e tests after the build has been created, although we [only have one](https://github.com/RaschidJFR/parse-dashboard/tree/bugfix/fix_lint_and_tests/src/lib/tests/e2e) for now.

But the rest of the tests are unit tests. From my short time spent in this repo, I perceive that they are loading components directly from the folder `src/`, so it doesn't make any difference to have the build created in the output folder. Although it wouldn't hurt to add this step in the job matrix, I think it makes more sense to add it during or before the build step.

And then the lining. We need to run it before Jest because unit tests fail if imports use double quotes instead of single quotes.

As I'm not very familiar with the github workflows, I'll trust your advice on where to add each of these 3 tasks (linting, unit tests, e2e).

mtrezza replied:
> loading components directly from the folder src/

I think we could improve that by loading from the built resource, to verify the build; we also do that in Parse Server repo tests where we load components from /lib instead of /src folder. That doesn't have to be part of this PR, but we can keep that in mind.

> And then the lining. We need to run it before Jest

The jobs are not run sequentially, they run in parallel as it is currently set up. Lint compliance should not have any code impact, just as lint-fixing should not have any code impact, it's merely a code style refactor.

> unit tests fail if imports use double quotes instead of single quotes.

That's strange where did you notice that?

> where to add each of these 3 tasks (linting, unit tests, e2e)

I'd suggest:
- re-add the commented out `check-lint` job where it originally was in the workflow file, just uncomment it and remove lint from `pretest` in `package.json`
- remove the `check-jest` job you added in this PR and just add a step to the existing `check-build` job:
  ```
   - name: Tests
      run: npm test
  ```
  this will build and test in all 3 node versions separately, which will hopefully make sense as we intent to add more UI tests, see https://github.com/parse-community/parse-dashboard/issues/2007. I'm not too familiar with react UI tests, but I assume they should run in each node version (and possibly multiple web render engines) just like the bundle building.

mtrezza commented on package.json:
```diff
@@ -88,6 +88,7 @@
     "@semantic-release/github": "7.2.3",
     "@semantic-release/npm": "7.1.3",
     "@semantic-release/release-notes-generator": "9.0.3",
+    "@types/jest": "^27.4.0",
```
Could you remove the range op from the version number?

RaschidJFR committed:
Merge branch 'alpha' into bugfix/fix_lint_and_tests

RaschidJFR commented:
I need a hand here with the CI step [Lint](https://github.com/parse-community/parse-dashboard/runs/4878290708?check_suite_focus=true) (I should rename it in case it stays). It is failing to import `.css` files. I believe it may be something related to babel. I'm not too familiar, so I'd appreciate pointing me in the right direction. Meanwhile, I'll start googling around.

mtrezza committed:
ci: bump environment

mtrezza committed:
bump node

mtrezza committed:
ci: bump environment (#2034)

KodeDave committed:
fix: adding internal class (e.g. `_User`) fails due to prefixed underscore (#2036)

semantic-release-bot committed:
chore(release): 4.0.0-alpha.16 [skip ci]

# [4.0.0-alpha.16](https://github.com/ParsePlatform/parse-dashboard/compare/4.0.0-alpha.15...4.0.0-alpha.16) (2022-02-10)

### Bug Fixes

* adding internal class (e.g. `_User`) fails due to prefixed underscore ([#2036](https://github.com/ParsePlatform/parse-dashboard/issues/2036)) ([f80bd07](https://github.com/ParsePlatform/parse-dashboard/commit/f80bd07a42b19fc4fa2632e0147fa72812a87c2f))

dependabot[bot] committed:
refactor: bump follow-redirects from 1.14.7 to 1.14.8 (#2039)

snyk-bot committed:
refactor: upgrade typescript from 4.5.4 to 4.5.5 (#2038)

dependabot[bot] committed:
refactor: bump node-sass from 6.0.1 to 7.0.0 (#2037)

snyk-bot committed:
refactor: upgrade otpauth from 7.0.9 to 7.0.10 (#2044)

snyk-bot committed:
refactor: upgrade graphql from 16.2.0 to 16.3.0 (#2043)

RaschidJFR committed:
ci: run unit tests in 'check-build'; readd 'check-lint'

RaschidJFR committed:
chore(jest): configure to map '.css' and avoid failing

RaschidJFR committed:
Merge branch 'alpha' into bugfix/fix_lint_and_tests

snyk-bot committed:
fix: security upgrade prismjs from 1.26.0 to 1.27.0 (#2047)

semantic-release-bot committed:
chore(release): 4.0.0-alpha.17 [skip ci]

# [4.0.0-alpha.17](https://github.com/ParsePlatform/parse-dashboard/compare/4.0.0-alpha.16...4.0.0-alpha.17) (2022-02-23)

### Bug Fixes

* security upgrade prismjs from 1.26.0 to 1.27.0 ([#2047](https://github.com/ParsePlatform/parse-dashboard/issues/2047)) ([3afb24e](https://github.com/ParsePlatform/parse-dashboard/commit/3afb24e708a69560732a725574953333431c1ca9))

snyk-bot committed:
refactor: upgrade @babel/runtime from 7.16.7 to 7.17.0 (#2048)

snyk-bot committed:
fix: upgrade @babel/runtime from 7.17.0 to 7.17.2 (#2055)

Snyk has created this PR to upgrade @babel/runtime from 7.17.0 to 7.17.2.

See this package in npm:
https://www.npmjs.com/package/@babel/runtime

See this project in Snyk:
https://app.snyk.io/org/acinader/project/3e039b91-2450-4b56-8420-baf56cab388e?utm_source=github&utm_medium=referral&page=upgrade-pr

semantic-release-bot committed:
chore(release): 4.0.0-alpha.18 [skip ci]

# [4.0.0-alpha.18](https://github.com/ParsePlatform/parse-dashboard/compare/4.0.0-alpha.17...4.0.0-alpha.18) (2022-03-02)

### Bug Fixes

* upgrade @babel/runtime from 7.17.0 to 7.17.2 ([#2055](https://github.com/ParsePlatform/parse-dashboard/issues/2055)) ([93335e9](https://github.com/ParsePlatform/parse-dashboard/commit/93335e9ab52e7b1f203b7939abde5c12b42dd2a7))

snyk-bot committed:
refactor: upgrade body-parser from 1.19.1 to 1.19.2 (#2057)

snyk-bot committed:
fix: upgrade express from 4.17.2 to 4.17.3 (#2058)

Snyk has created this PR to upgrade express from 4.17.2 to 4.17.3.

See this package in npm:
https://www.npmjs.com/package/express

See this project in Snyk:
https://app.snyk.io/org/acinader/project/3e039b91-2450-4b56-8420-baf56cab388e?utm_source=github&utm_medium=referral&page=upgrade-pr

Co-authored-by: Manuel <5673677+mtrezza@users.noreply.github.com>

semantic-release-bot committed:
chore(release): 4.0.0-alpha.19 [skip ci]

# [4.0.0-alpha.19](https://github.com/ParsePlatform/parse-dashboard/compare/4.0.0-alpha.18...4.0.0-alpha.19) (2022-03-10)

### Bug Fixes

* upgrade express from 4.17.2 to 4.17.3 ([#2058](https://github.com/ParsePlatform/parse-dashboard/issues/2058)) ([f8dc602](https://github.com/ParsePlatform/parse-dashboard/commit/f8dc6027f57143d72083f593455b23e304de934a))

RaschidJFR committed:
test(parse-sdk): add e2e test case for loading sdk

RaschidJFR committed:
test(jest): run with flag `--unhandled-rejections`

RaschidJFR commented:
@mtrezza It looks like the parse SDK [cannot be loaded](https://github.com/parse-community/parse-dashboard/pull/2004/files#diff-0e89fb988c29a42c6eba129edb05f47335370d1a76c2e837d2d8e63c40324b60) by Jest in Node v16, causing several tests to fail. However, this doesn't seem related to the SDK package itself (all tests passing OK in v16)

So far, the only workaround I found is to use the flag `--unhandled-rejections` in [package.json](https://github.com/parse-community/parse-dashboard/pull/2004/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519)  (see https://developer.ibm.com/blogs/nodejs-15-release-blog/).

Let me know your thoughts. If you find it ok, we're ready to merge.

mtrezza commented:
It's likely that we have some unhandled rejections, I remember we have (had?) this issue in Parse Server as well.

mtrezza committed:
Merge branch 'alpha' into bugfix/fix_lint_and_tests

mtrezza commented:
The only issue I see with adding the flag is that it may obscure issues with unhandled rejections. Especially since we are running tests, I am not sure whether it makes sense to run them in a non-representative environment. We may see tests pass, but dashboard won't actually run in certain Node environments.

How about the Parse JS SDK, in which Node versions do its CI tests run?

RaschidJFR commented:
I think you’re right. The SDK CI is running on v14. Should we drop support for v16 until the SDK officially supports it?

mtrezza commented:
I think that would be the process. Do you want to open a PR in the JS SDK and add Node 16 to the CI? Once it runs on Node 16, we could continue here without any errors hopefully.

RaschidJFR committed:
test(*): mock module 'idb-keyval' to support tests in Node v16

RaschidJFR committed:
test(jest): mock module 'idb-keyval' locally instead of globally

RaschidJFR commented:
Finally fixed. No need to modify the SDK repo. It was enough to [mock](https://github.com/parse-community/parse-dashboard/pull/2004/files#diff-c7cd0fbb75520d5427a4fc10731edb8e1d65f26ed88f991bc084c3a8b1446fdc) the module `idb-keyval` [from the SDK](https://github.com/parse-community/Parse-SDK-JS/blob/alpha/src/IndexedDBStorageController.js#L5).

mtrezza commented:
Could you resolve the conflicts?

mtrezza committed:
Merge branch 'alpha' into bugfix/fix_lint_and_tests

mtrezza commented:
I hope I resolved the conflicts correctly in package-lock. Let's see whether it passes.

mtrezza commented:
@RaschidJFR Could you take a last look before merge, since this has been stale for some days.

Note to myself:
- [x] add Lint CI job as required after merging

RaschidJFR commented:
LGTM. Thanks @mtrezza

parseplatformorg commented:
🎉 This change has been released in version [4.1.0-alpha.2](https://github.com/parse-community/parse-dashboard/releases/tag/4.1.0-alpha.2)
